### PR TITLE
Allow multiple TAR streams when Building Image

### PR DESCRIPTION
**Pull request by bedla**
Hi, 
I found that I am unable to build image with multiple tar streams. It is because calling `client.buildImageCmd(tarInputStream).withDockerfile(file)` calls two times method `withTarInputStream` with result that only last one is used (first is overwriten by second).
My use-case is: I have to use `withDockerfile(..)` method with path to custom Dockerfile also sending custom files to daemon necessary for build.
Please check PR,
Ivos

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/1301)

### CODE REVIEW

**Comment by bedla** [comment]
- Travis-ci - failing with https://github.com/docker-java/docker-java/issues/698 (investigating fix)
- Github CI - JVM crashed

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java [hunk 1]
```diff
@@ -214,7 +214,7 @@
     BuildImageCmd withNetworkMode(String networkMode);
 
     /**
-     *@since {@link RemoteApiVersion#VERSION_1_32}
+     * @since {@link RemoteApiVersion#VERSION_1_32}
```
Comment (bsideup) [inline docker-java-api/src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java]: please avoid unrelated changes

File: docker-java-core/src/main/java/com/github/dockerjava/core/async/JsonStreamProcessor.java [hunk 2]
```diff
@@ -39,7 +39,8 @@ public JsonStreamProcessor(Class<T> clazz) {
 
     public JsonStreamProcessor(ObjectMapper objectMapper, TypeReference<T> typeReference) {
         this.typeReference = typeReference;
-        this.objectMapper = objectMapper.copy().enable(JsonParser.Feature.AUTO_CLOSE_SOURCE);
+        this.objectMapper = objectMapper.copy()
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/async/JsonStreamProcessor.java]: please avoid unrelated changes

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java [hunk 3]
```diff
@@ -355,7 +363,7 @@ public BuildImageCmd withDockerfilePath(String dockerfilePath) {
     @Override
     public BuildImageCmdImpl withTarInputStream(InputStream tarInputStream) {
         checkNotNull(tarInputStream, "tarInputStream is null");
-        this.tarInputStream = tarInputStream;
+        this.tarInputStreams.add(tarInputStream);
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: behaviour change: previous implementation was overwriting it, while the new one only adds. Could you please restore the previous one?

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java [hunk 4]
```diff
@@ -405,12 +413,25 @@ public BuildImageCmd withTarget(String target) {
...
+            try {
+                tarInputStream.close();
+            } catch (IOException e) {
+                exceptions.add(e);
+            }
         }
-    }
 
+        if (!exceptions.isEmpty()) {
+            IllegalStateException e = new IllegalStateException();
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: please use `UncheckedIOException`

File: docker-java/src/test/java/com/github/dockerjava/cmd/CmdIT.java [hunk 5]
```diff
@@ -71,6 +71,6 @@ public FactoryType getFactoryType() {
     }
 
     @Rule
-    public DockerRule dockerRule = new DockerRule( this);
+    public DockerRule dockerRule = new DockerRule(this);
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/CmdIT.java]: unrelated change

**Comment by bsideup** [comment]
@bedla half way thought reviewing the PR and I have a feeling that it shouldn't be handled in the `BuildImageCmd`. If you need to provide multiple tar streams, why not using the utility method that will glue them together? (like the one you introduced in the PR)
That said, we could still add that method (although I think it should not write it to the file system but iterate entries of the provided streams)

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java (hunk 1 above)
Comment (bedla) [reply docker-java-api/src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java]: ok

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java (hunk 4 above)
Comment (bedla) [reply docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: ok

**Comment by bedla** [comment]
thx @bsideup for review. 
main theme of this PR is that when client code calls `withDockerfile` it loads `Dockerfile` and calls `withTarInputStream`. when in same run client code calls `withTarInputStream` previous tar-stream with `Dockerfile` is lost without warning (and it is hard to find why docker daemon is complaining about missing files).
my thought was that we have two possibilities how to fix this glitch
- first validate `tarInputStream` to having `null` when calling `withTarInputStream` and throwing exception if not. 
- second allow multiple tar streams, and merging them together (like I did in PR).
I understand that usage of utility tar-merging method would be nicer, but my code dont have chance to know (from current api) what stream is already assigned to build command.
what do you think?

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java (hunk 3 above)
Comment (bedla) [reply docker-java-core/src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: see comment below.

**Comment by bsideup** [comment]
@bedla 
There is also option number 3: instead of calling `withTarInputStream` from `withDockerfile`, store the path and decide how to create the `InputStream` later.

This sounds like a better fix for the issue (that should be defined as "`withDockerfile` overrides `InputStream` that was previously set `withTarInputStream` which is unexpected")

**Comment by bedla** [comment]
ok, I will investigate option 3. 
anyways IMHO option 1 is still valid, right?

**Comment by bsideup** [comment]
@bedla it is, more or less. Although, since the commands are builders, usually the validation is done at build (in our case, `exec`) time

**Comment by bedla** [comment]
@bsideup ok, I have invetigated it and have following thoughts.
- it is ok to do all the stuff at build/validation/`exec()` time
- but when both `withDockerfile` and `withTarInputStream` methods were called I have to do something:
1. do nothing (like curent behavior) and silently override one of generated `InputStreams`
2. make `InputStream` from `withDockerFile` higher priority and silently (log warning?) ignore second one
3. make `InputStream` from `withTarInputStream` higher priority and silently (log warning?) ignore second one
4. in case of both methods had set `InputStream` throw validation exception (opposed to point 2 and 3)
5. automatically merge both Tar `InputStreams` (like I did in utility method from PR)
6. manually merge both Tar `InputStreams` with new "tar-inputstream-merger" (so let client calling it decide). In this case I have to add method like `BuildImageCmd#withTarInputStreamMerger` (eg with signature `BiFunction<InputStream, InputStream, InputStream> merger`)

What do you think? From my point of view (and original thought) number 5 is the best. No. 6 is kind of ok. No. 2, 3, and 4 are too invasive, imo. And No. 1 I dont like at all.

thx,
Ivos

**Comment by bsideup** [comment]
@bedla after some consideration, it was decided that `docker-java` library will provide a low level, 1:1 mapping to the Docker API. Some helpers will remain, but, overall, the API should follow Docker's.

Given this information, I am not sure we should be doing this PR, since your use case can be achieved by providing a merged Tar stream (without using the `withDockerfile` method). WDYT?
